crates/shared/src/oneinch_api.rs

```diff
@@ -384,6 +386,60 @@ where
     serde_json::from_str(&response?).context("1inch result parsing failed")
 }
 
+#[derive(Debug, Clone)]
+pub struct ProtocolCache(Arc<Mutex<TimedSizedCache<(), Vec<String>>>>);
+
+impl ProtocolCache {
+    pub fn new(cache_validity_in_seconds: u64) -> Self {
+        Self(Arc::new(Mutex::new(
+            TimedSizedCache::with_size_and_lifespan_and_refresh(
+                1,
+                cache_validity_in_seconds,
+                false,
+            ),
+        )))
+    }
+
+    pub async fn get_all_protocols(&self, api: &dyn OneInchClient) -> Result<Vec<String>> {
+        if let Some(cached) = self.0.lock().unwrap().cache_get(&()) {
+            return Ok(cached.clone());
+        }
+
+        let all_protocols = api.get_protocols().await?.protocols;
+        // In the mean time the cache could have already been populated with new protocols,
+        // which we would now overwrite. This is fine.
+        self.0.lock().unwrap().cache_set((), all_protocols.clone());
+
+        Ok(all_protocols)
+    }
+
+    pub async fn get_allowed_protocols(
+        &self,
+        disabled_protocols: &[String],
+        api: &dyn OneInchClient,
+    ) -> Result<Option<Vec<String>>> {
+        if disabled_protocols.is_empty() {
+            return Ok(None);
+        }
+
+        let allowed_protocols = self
+            .get_all_protocols(api)
+            .await?
+            .into_iter()
+            // linear search through the slice is okay because it's very small
+            .filter(|protocol| !disabled_protocols.contains(protocol))
+            .collect();
+
+        Ok(Some(allowed_protocols))
+    }
+}
+
+impl Default for ProtocolCache {
+    fn default() -> Self {
+        Self::new(60)
```

Comment:
Nit: We can make this a lot longer than 60 seconds. I don't expect them to add protocols often and even if they do it wouldn't be a problem to not use a new protocol right that moment. Could be an hour. On the other hand fetching the protocols isn't expensive either so not sure.